File and FileReader (#325)
javascript-tutorial/es.javascript.info: 4-binary/04-file/article.md

> # Archivo y Lector de Archivos (File and FileReader en inglés)

Reviewer: ```suggestion
# File y FileReader
```

> Un [Archivo](https://www.w3.org/TR/FileAPI/#dfn-file) hereda de  `Blob` y es extendido con capacidades relacionadas con el sistema de archivos.

Reviewer: ```suggestion
Un [File](https://www.w3.org/TR/FileAPI/#dfn-file) hereda de `Blob` y extiende las capacidades relacionadas con el sistema de archivos.
```

Hay dos maneras de obtenerlo

Primero, hay un constructor, similar al de `Blob`:

```js

> new File(fileParts, fileName, [options])

Reviewer: ```suggestion
new File(partesDelArchivo, nombreDelArchivo, [opciones])
```

> new File(partesDelArchivo, nombreDelArchivo, [opciones])

Reviewer 2: Para mantener la consitencia con las líneas más abajo donde se mantuvieron los argumentos en inglés. 

```suggestion
new File(fileParts, fileName, [options])
```

```

> - **`fileParts`** -- es un array de valores Blob/BufferSource/String.

Reviewer 2: Yo lo dejaría como array, pero añadiendo "de tipo"

```suggestion
- **`fileParts`** -- es un array con valores de tipo Blob/BufferSource/String.
```

- **`fileName`** -- el nombre del archivo..
- **`options`** -- objeto opcional:

>     - **`lastModified`** -- la  marca de tiempo (fecha en mili-segundos) de la última modificación.

Reviewer 2: ```suggestion
    - **`lastModified`** -- la  marca de tiempo (fecha en mili-segundos, de tipo integer) de la última modificación.
```

Reviewer: @vplentinax en todo caso seria **de tipo entero** en lugar de de tipo integer

Reviewer 2: > @vplentinax en todo caso seria **de tipo entero** en lugar de **de tipo integer**

Cierto

> Segundo, a menudo obetenemos un archivo mediante un `<input type="file">` o un drag'n'drop, u otra interfaz del navegador. En este caso el archivo obtiene la información del Sistema Operativo.

Reviewer 2: Con este qué hacemos...

> Segundo, a menudo obetenemos un archivo mediante un `<input type="file">` o un drag'n'drop, u otra interfaz del navegador. En este caso el archivo obtiene la información del Sistema Operativo.

Reviewer: No lo se :p

```

```smart

> El input puede seleccionar varios archivos, por lo que `input.files` es una lista de dichos archivos . En este caso tenemos un solo archivo por lo que solo es necesario usar `input.files[0]`.

Reviewer 2: ```suggestion
El input puede seleccionar varios archivos, por lo que `input.files` es un array de dichos archivos . En este caso tenemos un solo archivo por lo que solo es necesario usar `input.files[0]`.
```

Reviewer: @vplentinax yo no usaría array en todo caso usaría arreglo, pero me parece que es mejor usar lista.

Reviewer 2: > @vplentinax yo no usaría array en todo caso usaría arreglo, pero me parece que es mejor usar lista.

He usado array en todas las traducciones. Pero es elección de traductor.

Author: Originalmente habia usado el termino array, lo he visto mucho en docs y libros en español, ademas de que mis profesores en la universidad suelen usarlo. Creo que hay cosas que es mejor dejar en ingles. 
Proveer una traduccion, hace el contenido accesible, pero acceptemoslo,
en esta industria se habla inglés así que motivar al lector a ir familiarizandose siempre me parece una buena idea

```

> ## Lector de Archivos 

Reviewer 2: ```suggestion
## FileReader
```

> [Lector de Archivos](https://www.w3.org/TR/FileAPI/#dfn-filereader) es un objeto con el único porpósito de leer datos desde  objetos  de tipo `Blob` (y entonces `File` también).

Reviewer 2: ```suggestion
[FileReader](https://www.w3.org/TR/FileAPI/#dfn-filereader) es un objeto con el único porpósito de leer datos desde  objetos  de tipo `Blob` (por lo tanto `File` también).
```

El entrega los datos usando evetos, debido a que leerlos desde el disco puede tomar tiempo.

El constructor:

```js
let reader = new FileReader(); // sin argumentos
```

Los métodos principales:

- **`readAsArrayBuffer(blob)`** -- lee los datos en formato binario `ArrayBuffer`.
- **`readAsText(blob, [codificación])`** -- lee los datos como una cadena de texto con la codificación dada (por defecto es `utf-8`).
- **`readAsDataURL(blob)`** -- lee los datos binarios y los codifica como [Datos URIs] en base 64 (https://developer.mozilla.org/es/docs/Web/HTTP/Basics_of_HTTP/Datos_URIs).

- **`abort()`** -- cancela la operación.

> La opción de `read*` método depende de que formato preferimos y como vamos a usar los datos.

Reviewer 2: ```suggestion
La opción del método `read*` depende de qué formato preferimos y cómo vamos a usar los datos.
```

> - `readAsArrayBuffer` -- para archivos binarios, para hacer operaciones binarias de bajo nivel. Para operaciones de alto nivel, como slicing, `File` hereda de `Blob`, entonces podemos llamarlas directamente, sin tener que leer.

Reviewer: ```suggestion
- `readAsArrayBuffer` -- para archivos binarios, en donde se hacen operaciones binarias de bajo nivel. Para operaciones de alto nivel, como slicing, `File` hereda de `Blob` por lo que podemos llamarlas directamente sin tener que leer.
```

- `readAsText` -- para archivos de texto, cuando nesecitamos obtener una cadena.

> - `readAsDataURL` -- cuando nesecitamos usar estos datos valores de `src` en `img` o otras etiquetas html. Hay otra alternativa para leer archivos con este último fin, como es discutido en el capítulo <info:blob>: `URL.createObjectURL(file)`.

Reviewer: ```suggestion
- `readAsDataURL` -- cuando necesitamos usar estos datos valores de `src` en `img` u otras etiquetas html. Hay otra alternativa para leer archivos para eso como discutimos en el capítulo <info:blob>: `URL.createObjectURL(file)`.
```

> - `readAsDataURL` -- cuando necesitamos usar estos datos como valores de `src` en `img` u otras etiquetas html. Hay otra alternativa para leer archivos para eso como discutimos en el capítulo <info:blob>: `URL.createObjectURL(file)`.

Reviewer 2: ```suggestion
- `readAsDataURL` -- cuando necesitamos usar estos datos como valores de `src` en `img` u otras etiquetas html. Hay otra alternativa para leer archivos de ese tipo como discutimos en el capítulo <info:blob>: `URL.createObjectURL(file)`.
```

> En la medida que la lectura procede, suceden varios eventos:

Reviewer 2: ```suggestion
Mientras se va realizando la lectura, suceden varios eventos:
```

Author: @vplentinax 
Crees que no  se entiende bien de la manera que está ?

Reviewer 2: > @vplentinax
> Crees que no se entiende bien de la manera que está ?

Lo noto un tanto extraño, sí.

> - `loadstart` -- lectura iniciada.

Reviewer: ```suggestion
- `loadstart` -- la carga comenzó.
```

- `progress` -- ocurre mientras se lee.
- `load` -- lectura completada, sin errores.
- `abort` -- `abort()` ha sido llamado.
- `error` -- ha ocurrido un error .

> - `loadend` -- la lectura finalizó exitosa o no .

Reviewer 2: ```suggestion
- `loadend` -- la lectura finalizó, correctamente o, por el contrario, con errores .
```

Reviewer: ```suggestion
- `loadend` -- la lectura finalizó de manera exitosa o con errores .
```

> - `loadend` -- lectura termindada tanto exitosa como fallidamente.

Reviewer: ```suggestion
- `loadend` -- la lectura finalizó exitosa o no .
```

</script>
```

```smart header="`FileReader` para blobs"

> Como mencionamos en el capítulo <info:blob>, `FileReader` no solo lee archivos, sino también cualquier blob.

Reviewer: ```suggestion
Como mencionamos en el capítulo <info:blob>, `FileReader` no solo lee archivos sino también cualquier blob.
```

Podemos usarlo para convertir un blob a otro formato:
- `readAsArrayBuffer(blob)` -- a `ArrayBuffer`,
- `readAsText(blob, [encoding])` -- a cadena (una alternativa al `TextDecoder`),
- `readAsDataURL(blob)` -- a Dato Uri.
```

```smart header="`FileReaderSync` está disponible dentro de  Web Workers"
Para Web Workers, también existe una variante síncrona de `FileReader`, llamada [FileReaderSync](https://www.w3.org/TR/FileAPI/#FileReaderSync).

Sus metodos`read*` no generan eventos, sino que devuelven un resultado, como las funciones regulares.

Esto es solo dentro de un Web Worker, debido a que  demoras en  llamadas síncronas  mientras se lee el archivo en Web Worker no son tan importantes. No afectan la página.

```

> ## Sumario

Reviewer: ```suggestion
## Resumen
```

Los objetos `File` heredan de  `Blob`.

En adición a los métodos y porpiedades de `Blob`, los objetos `File` también tienen `name` and `lastModified` mas la habilidad interna de leer del sistema de archivos. Usualmente obtenemos los objetos `File` mediante la entrada por el usuario con `<input>` o un evento Drag'n'Drop (`ondragend`).

Objetos `FileReader` pueden leer desde un archivo o un blob, en uno de estos tres formatos:
- String (`readAsText`) .
- `ArrayBuffer` (`readAsArrayBuffer`).
- Datos Uri (`readAsDataURL`).

En muchos casos, no nesecitamos leer el contenido de un archivo. Como hicimos con blobs, podemos crear 

> un url corto con `URL.createObjectURL(file)` y asignárselo a un `<a>` o `<img>`. De esta manera el archivo puede ser descargado o ser mostrado como una imagen, o como parte de un canvas etc.

Reviewer 2: Sugerencia de @danilobrinu que se sobreescribió

```suggestion
un enlace corto con `URL.createObjectURL(file)` y asignárselo a un `<a>` o `<img>`. De esta manera el archivo puede ser descargado o ser mostrado como una imagen o como parte de un canvas, etc.
```

> Y si vamos a mandar un `File` por la red, es también fácil: APIs como `XMLHttpRequest` o `fetch` acceptan nativamente objetos `File` . 

Reviewer 2: Hice una pequeña edición a tu sugerencia. @danilobrinu

</script>
```

```smart header="`FileReader` para blobs"
Como mencionamos en el capítulo <info:blob>, `FileReader` no solo lee archivos sino también cualquier blob.

Podemos usarlo para convertir un blob a otro formato:
- `readAsArrayBuffer(blob)` -- a `ArrayBuffer`,
- `readAsText(blob, [encoding])` -- a cadena (una alternativa al `TextDecoder`),
- `readAsDataURL(blob)` -- a Dato Uri.
```

```smart header="`FileReaderSync` está disponible dentro de  Web Workers"

> Para Web Workers, también existe una variante síncrona de `FileReader`, llamada [FileReaderSync](https://www.w3.org/TR/FileAPI/#FileReaderSync).

Reviewer: ```suggestion
Para los Web Workers también existe una variante síncrona de `FileReader` llamada [FileReaderSync](https://www.w3.org/TR/FileAPI/#FileReaderSync).
```

> Sus metodos`read*` no generan eventos, sino que devuelven un resultado, como las funciones regulares.

Reviewer: ```suggestion
Sus metodos `read*` no generan eventos sino que devuelven un resultado como las funciones regulares.
```

Esto es solo dentro de un Web Worker, debido a que  demoras en  llamadas síncronas  mientras se lee el archivo en Web Worker no son tan importantes. No afectan la página.

```

## Resumen

Los objetos `File` heredan de  `Blob`.

> En adición a los métodos y porpiedades de `Blob`, los objetos `File` también tienen `name` and `lastModified` mas la habilidad interna de leer del sistema de archivos. Usualmente obtenemos los objetos `File` mediante la entrada por el usuario con `<input>` o un evento Drag'n'Drop (`ondragend`).

Reviewer 2: ```suggestion
En adición a los métodos y porpiedades de `Blob`, los objetos `File` también tienen `name` y `lastModified` mas la habilidad interna de leer del sistema de archivos. Usualmente obtenemos los objetos `File` mediante la entrada por el usuario con `<input>` o un evento arrastrar y soltar(Drag'n'Drop).
```

Reviewer: ```suggestion
En adición a los métodos y porpiedades de `Blob`, los objetos `File` también tienen `name` and `lastModified` mas la habilidad interna de leer del sistema de archivos. Usualmente obtenemos los objetos `File` mediante la entrada por el usuario con `<input>` o un evento arrastar y soltar (`ondragend`).
```

Reviewer 2: El "and" debe traducirse.
